### ISSUE TITLE
Ability to split and parallelize test runs

### ISSUE THREAD

**Issue by toots** [ocaml/dune#6179]
Applogize if this has already been discussed but I couldn't find any opem request for this.

This request is related to https://github.com/ocaml/dune/issues/6142: our test suite is getting big and takes longer to execute. With most test suite, there is the ability to assign a subset of the tests to a given instance, making it possible to parallelize test runs accros multiple runners.

I was not able to find this option with `dune`.

**Comment by emillon** [comment]
Indeed, there's no way to do that at the moment. I would say that for now the preferred way is to scale up (run on a larger machine) rather scale out (on several machines). One thing that will help this is making the dune cache distributed between machines, so that runners can share a common cache. The machine-local part already exists but the distributed part is still being worked out.

**Comment by rgrinberg** [comment]
What do you think of marking a subset of your tests with a specific alias? Would that work for you?

**Comment by toots** [comment]
> What do you think of marking a subset of your tests with a specific alias? Would that work for you?

Yes, this works. I just implemented it. It fits out workflow just fine b/c we're already auto-generating our dune files for this. However, this ties up our dune definitions to the CI, which is not ideal. That's another case where having more support for dynamic targets would be great.

Happy to have a work-around though!

**Comment by rgrinberg** [comment]
In yesterday's meeting we discussed the concept of "forkers" that we'd like to add to dune. Forkers would allow you to execute actions on different instances. The purpose of forkers is not necessarily performance, but it might help here anyway. There might be a mechanism for assigning subsets of targets to specific forkers. I'll let @snowleopard expand if there might be something that might help here.

**Comment by snowleopard** [comment]
To clarify, a "forker" is a process that Dune can communicate with via Dune RPC to execute build actions. Let's refer to them as "Dune action runners" for clarity. Separate action runners are useful in the context of action isolation: we might want to restrict some actions from seeing the whole source tree in order to provide some information flow guarantees. To do that, we can wrap Dune action runners in different `bwrap` instances (or use some other process isolation mechanism).

For this to work well, I think action runners should be on the same file system as Dune, as otherwise there will be quite a lot of copying of action dependencies. Having said that, maybe for long running tests it would be worth sending them over the network to another machine.

I've just created an issue about Dune action runners: https://github.com/ocaml/dune/issues/6199
